db/src/timelines.rs

```diff
@@ -0,0 +1,442 @@
+    conn.execute(&format!(
+        "UPDATE timelined_transactions SET timeline = {} WHERE tx IN ({})",
+            new_timeline,
+            ::repeat_values(tx_ids.len(), 1)
+        ), &(tx_ids.iter().map(|x| x as &rusqlite::types::ToSql).collect::<Vec<_>>())
+    )?;
+    Ok(())
+}
+
+/// Get terms for tx_id, reversing them in meaning (swap add & retract).
+fn reversed_terms_for(conn: &rusqlite::Connection, tx_id: Entid) -> Result<Vec<TermWithoutTempIds>> {
+    let mut stmt = conn.prepare("SELECT e, a, v, value_type_tag, tx, added FROM timelined_transactions WHERE tx = ? AND timeline = ? ORDER BY tx DESC")?;
+    let mut rows = stmt.query_and_then(&[&tx_id, &MAIN_TIMELINE], |row| -> Result<TermWithoutTempIds> {
+        let op = match row.get_checked(5)? {
+            true => OpType::Retract,
+            false => OpType::Add
+        };
+        Ok(Term::AddOrRetract(
+            op,
+            KnownEntid(row.get_checked(0)?),
+            row.get_checked(1)?,
+            TypedValue::from_sql_value_pair(row.get_checked(2)?, row.get_checked(3)?)?,
+        ))
+    })?;
+
+    let mut terms = vec![];
+
+    while let Some(row) = rows.next() {
+        terms.push(row?);
+    }
+    Ok(terms)
+}
+
+/// Move specified tail-end transactions off of main timeline.
+/// 
+/// It's not possible to determine a high water-mark of partitions
+/// when rewinding transactions just by inspection of their datoms.
+/// We depend on our caller to supply the correct "target" partition map.
+pub fn move_from_main_timeline(conn: &rusqlite::Connection, schema: &Schema,
+    partition_map: PartitionMap, tx_ids: &[Entid], new_timeline: Entid) -> Result<(Option<Schema>, PartitionMap)> {
+
+    if new_timeline == MAIN_TIMELINE {
+        bail!(DbErrorKind::NotYetImplemented(format!("Can't move transactions to main timeline")));
+    }
+
+    let mut sorted_tx_ids = tx_ids.to_vec();
+    sorted_tx_ids.sort_by(|t1, t2| t1.cmp(t2));
+
+    ensure_on_tail_of(conn, &sorted_tx_ids, MAIN_TIMELINE)?;
+
+    let mut last_schema = None;
+    for tx_id in sorted_tx_ids {
+        let reversed_terms = reversed_terms_for(conn, tx_id)?;
+
+        // Rewind schema and datoms.
+        let (_, _, new_schema, _) = transact_terms_with_action(
+            conn, partition_map.clone(), schema, schema, NullWatcher(),
+            reversed_terms.into_iter().map(|t| t.rewrap()),
+            InternSet::new(), TransactorAction::Materialize
+        )?;
+        last_schema = new_schema;
+    }
+
+    // Move transactions over to the target timeline.
+    move_transactions_to(conn, tx_ids, new_timeline)?;
+
+    Ok((last_schema, db::read_partition_map(conn)?))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use edn;
+
+    use std::borrow::{
+        Borrow,
+    };
+
+    use debug::{
+        TestConn,
+    };
+
+    use bootstrap;
+
+    // TODO I don't think "moving timelines" belongs in a conn (which would handle this),

```

Comment:
This is exactly the kind of thing that would live in `TestConn`.  You want rewinding to do the right thing, schema and partition map and all.  (I'm not sure it would live in top-level `Conn`, since it's Tolstoy specific.  Maybe.)

Reply:
I didn't want to hide the nature of this function behind a test interface. E.g. it explicitly _doesn't_ update a conn (which would live layers above, anyway...), so being explicit about it in tests seemed useful. But, having this stuff in `TestConn` is handy.

This is an interesting question, it touches on how we split stuff across crates (e.g. `TestConn` present in `db`...)